In [86]:
# Importing necessary libraries
import pandas as pd
from myfunctions import set_importer, scores, data_preparation, data_cleaner
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.metrics import plot_confusion_matrix, train_test_split
from sklearn.ensemble import RandomForestClassifier

sns.set_style("whitegrid")
%matplotlib inline

# setting font size for all plots
font = {'size'   : 16}

plt.rc('font', **font)

# The best Model

After experimenting with various models, a random forest classifier was found to be the ideal model

In [75]:
class myRandomForestClassifier:
    def __init__(self):
        self.rf_classifier =  RandomForestClassifier(criterion='gini', random_state=21, min_samples_split=70, min_samples_leaf=100, n_estimators=300)
    
    def Xy_cleanprep(self, X, y):
        #1. data cleaning
        X_cleaned, y_cleaned = data_cleaner(X, y)

        # 2. data preparation
        X_ordinal_encoded, X_onehotencoded, y_prepared = data_preparation(X_cleaned, y_cleaned)

        columns_to_select = ['gps_height','longitude','latitude','region_code','district_code',
        'population','permit','construction_year','basin','extraction_type_class','payment','quality_group','quantity_group','source_class','waterpoint_type_group']

        return X_ordinal_encoded[columns_to_select], y_prepared
        
    
    def fit(self, X, y):
        
        X_ready, y_ready = self.Xy_cleanprep(X, y)

        return self.rf_classifier.fit(X_ready, y_ready)
    
    def predict(self, X_true):
        columns_to_select = ['gps_height','longitude','latitude','region_code','district_code',
        'population','permit','construction_year','basin','extraction_type_class','payment','quality_group','quantity_group','source_class','waterpoint_type_group']
        return self.rf_classifier.predict(X_true[columns_to_select])



In [76]:
# importing the data frame
df = set_importer('../analysis-dfs/df.csv')
df.head(10)

id  amount_tsh date_recorded                  funder  gps_height  \
0  67743         0.0    2013-01-28                  Unicef         263   
1  19816         0.0    2012-10-01                    Dwsp           0   
2  54551         0.0    2012-10-09                   Rwssp           0   
3  53934         0.0    2012-11-03                Wateraid           0   
4  48451       500.0    2011-07-04                  Unicef        1703   
5  58155         0.0    2011-09-04                  Unicef        1656   
6  34169         0.0    2011-07-22                  Hesawa        1162   
7  58500         0.0    2011-10-04                  Unicef        1510   
8  22308         0.0    2013-02-25  Government Of Tanzania        1273   
9  19685         0.0    2013-03-05  Government Of Tanzania        1443   

          installer  longitude   latitude              wpt_name  num_private  \
0            UNICEF  38.486161 -11.155298  Zahanati Ya Nanyumbu            0   
1              DWSP  33.362410  -3.766365            Kwa Ngomho            0   
2               DWE  32.620617  -4.226198          Tushirikiane            0   
3         Water Aid  32.711100  -5.146712     Kwa Ramadhan Musa            0   
4               DWE  34.642439  -9.106185       Kwa John Mtenzi            0   
5               DWE  34.569266  -9.085515       Kwa Rose Chaula            0   
6               DWE  32.920154  -1.947868                Ngomee            0   
7               DWE  34.586901  -8.980014               Shuleni            0   
8               DWE  37.422751  -3.317536             Kwa Tukai            0   
9  District council  37.611126  -3.263526          Kwa Kibakaya            0   

                     basin       subvillage       region  region_code  \
0  Ruvuma / Southern Coast       Mahakamani       Mtwara           90   
1                 Internal    Ishinabulandi    Shinyanga           17   
2          Lake Tanganyika  Nyawishi Center    Shinyanga           17   
3          Lake Tanganyika       Imalauduki       Tabora           14   
4                   Rufiji         Kidudumo       Iringa           11   
5                   Rufiji           Yeriko       Iringa           11   
6            Lake Victoria           Center       Mwanza           19   
7                   Rufiji        Mkanivega       Iringa           11   
8                  Pangani         Kitereni  Kilimanjaro            3   
9                  Pangani            Kiyao  Kilimanjaro            3   

   district_code              lga          ward  population public_meeting  \
0             63         Nanyumbu      Nanyumbu          58           True   
1              3  Shinyanga Rural        Samuye           0           True   
2              3           Kahama        Chambo           0           True   
3              6     Tabora Urban      Itetemia           0           True   
4              4           Njombe        Mdandu          35           True   
5              4           Njombe         Usuka          50           True   
6              1          Ukerewe      Ilangala        1000            NaN   
7              4           Njombe         Usuka         350           True   
8              4      Moshi Rural       Kimochi         140           True   
9              1            Rombo  Mengwe Manda           1           True   

               recorded_by scheme_management                  scheme_name  \
0  GeoData Consultants Ltd               VWC                          NaN   
1  GeoData Consultants Ltd               VWC                          NaN   
2  GeoData Consultants Ltd               NaN                          NaN   
3  GeoData Consultants Ltd               VWC                          NaN   
4  GeoData Consultants Ltd               WUA  wanging'ombe water supply s   
5  GeoData Consultants Ltd               WUA  wanging'ombe water supply s   
6  GeoData Consultants Ltd               NaN                          NaN   
7  GeoData Consultants Ltd     

In [77]:
X = df.drop('status_group', axis=1)
y = df.status_group
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.30, random_state=21)

In [78]:
clf = myRandomForestClassifier()
clf.fit(X_train, y_train)

RandomForestClassifier(min_samples_leaf=100, min_samples_split=70,
                       n_estimators=300, random_state=21)

In [79]:
# passing  the test set to a cleaner
X_test_ready, y_test_ready = clf.Xy_cleanprep(X_test, y_test)

In [85]:
y_test_rf_preds = clf.predict(X_test_ready)
# print('RandomForest Classifier best model results on the training set')
scores(y_test_ready, y_test_rf_preds)

The precision score is:	 0.8478688524590164
The recall score is:	 0.9998453448809156
The accuracy score is:	 0.8477576711250984
The f1 score is:	 0.9176069831807537


## Interpretating the final model

* Precision-Score - If the model assigns 1 then there is a 85% chance that a water pump is actually non functional.
* Recal-Score - If a given water pump is actually 1 (non-functional) then there is a 99.9% that this model will label it as 1 (non-functional), and a 0.1% chance that it will incorrectly label it as 0 (functional but need repair).
* Accuracy-Score - The model accrately assigns 84.8% to the correct label. (0.8375295043273013)